core/src/main/java/com/scalar/db/transaction/common/AbstractDistributedTransactionManager.java

```diff
@@ -177,8 +184,119 @@ public void rollback() throws RollbackException {
       }
     }
 
+    @Override
+    public DistributedTransaction getOriginalTransaction() {
+      if (transaction instanceof WrappedDistributedTransaction) {
+        return ((WrappedDistributedTransaction) transaction).getOriginalTransaction();
+      }
+      return transaction;
+    }
+  }
+
+  @VisibleForTesting
+  static class TransactionStateManagement extends AbstractDistributedTransaction
+      implements WrappedDistributedTransaction {
+
+    private enum Status {
+      ACTIVE,
+      COMMITTED,
+      COMMIT_FAILED,
+      ROLLED_BACK
+    }
+
+    private final DistributedTransaction transaction;
+    private Status status;
```

Comment:
It's better to make this `volatile` or use AtomicReference so that an update status is visible to other threads? For instance, a thread processing `commit()` method could fail to read the latest `ROLLED_BACK` status just updated by another thread that called `rollback()` method.

Reply:
Ah, ConsensusCommit and etc. which inherit AbstractDistributedTransaction have `@NotThreadSafe` annotation, so it's okay?